src/csharp/Grpc.Core/Internal/AsyncCallBase.cs

```diff
@@ -252,18 +259,39 @@ protected Exception TryDeserialize(byte[] payload, out TRead msg)
         /// </summary>
         protected void HandleSendFinished(bool success)
         {
+            bool delayCompletion = false;
             TaskCompletionSource<object> origTcs = null;
             lock (myLock)
             {
                 origTcs = streamingWriteTcs;
                 streamingWriteTcs = null;
 
+                if (!success && !finished && IsClient)
+                {
+                    // We should be setting this only once per call, following writes will be short circuited.
```

Comment:
Actually, good question. There was a problem with the implementation - because streamingWriteTcs was reset here, if there was a delayed write completion, one could have started another write even though the failed write hasn't finished yet from user's perspective (and that also meant that the statement in the comment was incorrect).

I changed the implementation to use a boolean flag isDelayedStreamingWriteCompletion and I'm leaving streamingWriteTcs set until the call actually finishes. Now everything should be alright hopefully. I also added a test that was exposing the issue with the implementation (it is passing now).

PTAL. Hopefully no more fixes will be needed.